chat-plugins/tickets.js

```diff
@@ -0,0 +1,562 @@
+				if (!ticket && ipTicket) ticket = ipTicket;
+				let helpRoom = Rooms(`help-${ticket.userid}`);
+				if (!helpRoom) {
+					// Should never happen
+					tickets[ticket.userid].open = false;
+					writeTickets();
+				} else {
+					if (!helpRoom.auth[user.userid]) helpRoom.auth[user.userid] = '+';
+					this.parse(`/join help-${ticket.userid}`);
+					return this.sendReply(`You already have an open ticket; please wait for global staff to respond.`);
+				}
+			}
+			if (Monitor.countTickets(user.latestIp)) return user.popup(`Due to high load, you are limited to creating ${Punishments.sharedIps.has(user.latestIp) ? `50` : `5`} tickets every hour.`);
+			if (!['PM Harassment', 'Battle Harassment', 'Chatroom Harassment', 'Inappropriate Content', 'Inappropriate Nickname', 'Inappropriate Pokemon Nicknames', 'Timerstalling', 'Global Staff Complaint', 'Appeal', 'IP-Appeal', 'ISP-Appeal', 'Lost Password', 'Report Last Ticket', 'Room Owner Complaint', 'Other'].includes(target)) return this.parse('/ticket');
+			let upper = false;
+			if (['Lost Password', 'Room Owner Complaint', 'Global Staff Complaint', 'Report Last Ticket'].includes(target)) upper = true;
+			if (target === 'Report Last Ticket') {
+				if (!ticket) return this.errorReply(`You can't report a ticket that dosen't exist.`);
+				target = `Report Last Ticket - ${ticket.type}`;
+			}
+			ticket = {
+				creator: user.name,
+				userid: user.userid,
+				open: true,
+				type: target,
+				created: Date.now(),
+				claimed: null,
+				escalated: upper,
+				ip: user.latestIp,
+			};
+			let helpRoom = Rooms(`help-${user.userid}`);
+			if (!helpRoom) {
+				helpRoom = Rooms.createChatRoom(`help-${user.userid}`, `[H] ${user.name}`, {
+					isPersonal: true,
+					isHelp: 'open',
+					isPrivate: 'hidden',
+					modjoin: (upper ? '&' : '%'),
+					auth: {[user.userid]: '+'},
+					introMessage: `<h2 style="margin-top:0">Help Ticket - ${user.name}</h2><p><b>Issue</b>: ${ticket.type}<br />${upper ? `An Upper` : `A Global`} Staff member will be with you shortly.</p>`,
+					staffMessage: `${upper ? `<p><h3>Do not post sensitive information in this room.</h3>Drivers and moderators can access this room's logs via the log viewer; please PM the user instead.</p>` : ``}<p><button class="button" name="send" value="/ticket close ${user.userid}">Close Ticket</button> <button class="button" name="send" value="/ticket escalate ${user.userid}">Escalate</button> ${upper ? `` : `<button class="button" name="send" value="/ticket escalate ${user.userid}, upperstaff">Escalate to Upper Staff</button>`}</p><p>To ban this user from creating tickets for two days, please use <code>/ticket ban ${user.userid}</code></p>`,
+				});
+				helpRoom.game = new HelpTicket(helpRoom, ticket);
+			} else {
+				helpRoom.isHelp = 'open';
+				if (helpRoom.expireTimer) clearTimeout(helpRoom.expireTimer);
+				if (upper && helpRoom.modjoin === '%') {
+					// Kick drivers and moderators out
```

Comment:
In practice I find this unnecessary. Let admins manually set modjoin if it's actually something that requires that much privacy.

Reply:
drivers/mods are only kicked out if a new ticket is created that has modjoin set to & by default, and a ticket room already exists with a lower modjoin setting.

Reply:
I don't know if that's the right approach...

Reply:
You could create a separate ticket room with the permissions set with a duplicate roomlog

Reply:
Could force the user to leave the old room and enter the new one to prevent confusion?

Reply:
No, I just straight up don't think any ticket room should be barred to staff. Any actual high-privacy interaction should probably be in PM.

Reply:
I spoke with Orda-Y about this a bit, one concern is locked users who need to communicate sensetive information cannot do so in PMs (once we implement that restriction). This also brings the concern about sensetive information in logs. Maybe we should allow locked users with an open escalated ticket to PM upper staff? (we can check this via the room's "game")